packages/block-editor/src/components/typewriter/index.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { Component, createRef } from '@wordpress/element';
+import { computeCaretRect, getScrollContainer } from '@wordpress/dom';
+import { withSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+
+class Typewriter extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.ref = createRef();
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.onMouseDown = this.onMouseDown.bind( this );
+		this.onTouchStart = this.onTouchStart.bind( this );
+		this.maintainCaretPositionOnSelectionChange = this.maintainCaretPositionOnSelectionChange.bind( this );
+		this.computeCaretRectOnSelectionChange = this.computeCaretRectOnSelectionChange.bind( this );
+		this.maintainCaretPosition = this.maintainCaretPosition.bind( this );
+		this.computeCaretRect = this.computeCaretRect.bind( this );
+		this.debouncedComputeCaretRect = debounce( this.computeCaretRect, 100 );
+		this.isSelectionEligibleForScroll = this.isSelectionEligibleForScroll.bind( this );
+	}
+
+	componentDidMount() {
+		window.addEventListener( 'scroll', this.debouncedComputeCaretRect, true );
+		window.addEventListener( 'resize', this.debouncedComputeCaretRect, true );
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'scroll', this.debouncedComputeCaretRect, true );
+		window.removeEventListener( 'resize', this.debouncedComputeCaretRect, true );
+		document.removeEventListener( 'selectionchange', this.maintainCaretPositionOnSelectionChange );
+		document.removeEventListener( 'selectionchange', this.computeCaretRectOnSelectionChange );
+		window.cancelAnimationFrame( this.rafId );
+		this.debouncedComputeCaretRect.cancel();
+	}
+
+	computeCaretRect() {
+		if ( this.isSelectionEligibleForScroll() ) {
+			this.caretRect = computeCaretRect();
+		}
+	}
+
+	maintainCaretPositionOnSelectionChange() {
+		document.removeEventListener( 'selectionchange', this.maintainCaretPositionOnSelectionChange );
+		this.maintainCaretPosition();
+	}
+
+	computeCaretRectOnSelectionChange() {
+		document.removeEventListener( 'selectionchange', this.computeCaretRectOnSelectionChange );
+		this.computeCaretRect();
+	}
+
+	isSelectionEligibleForScroll() {
+		return (
+			// There should be one and only one block selected.
+			this.props.selectedBlockClientId &&
+			// The component must contain the selection.
+			this.ref.current.contains( document.activeElement ) &&
+			// The active element must be contenteditable.
+			document.activeElement.getAttribute( 'contenteditable' )
+		);
+	}
+
+	maintainCaretPosition() {
+		if ( ! this.caretRect ) {
+			return this.computeCaretRect();
```

Comment:
Where is this return value used?

**Edit:** Okay, after thinking on it, I guess you're relying on `computeCaretRect` as a side-effect, and using `return` as a conciseness convenience. I'll let my initial confusion stand on its own as an anecdote in favor of the explicit:

```suggestion
			this.computeCaretRect();
			return;
```

Further, maybe `computeCaretRect` could be more explicit that it has the side-effect in being responsible to _assign_ a value, e.g. `assignCaretRect`.